---
title: 'Figures'
author: 'Christoph Hafemeister'
date: '`r format(Sys.time(), "%B %d, %Y %H:%M:%S %Z")`'
output:
  html_document:
    toc: TRUE
    toc_depth: 3
    toc_float: TRUE
    code_folding: "hide"
params:
  config_file: "config.yaml"
---

# Setup

```{r}
# load project-specific parameters
config <- yaml::read_yaml(params$config_file)

suppressPackageStartupMessages({
  library('Matrix')
  library('dplyr')
  library('Seurat')
  library('sctransform')
  library('reshape2')
  library('ggplot2')
  library('patchwork')
  library('ggrepel')
  source(file.path(config$project_root, 'R', 'utils.R'))
})

knitr::opts_chunk$set(comment = NA, fig.width = 5, fig.height = 3.5, out.width = '60%',
                      warning = TRUE, error = TRUE, echo = TRUE, message = TRUE,
                      dpi = 100)
options(dplyr.summarise.inform = FALSE)
options(tibble.width = Inf)
options(DT.options = list(paging = TRUE, searching = TRUE, autoWidth = TRUE))
options(ggrepel.max.overlaps = Inf)

old <- theme_set(theme_bw(base_size = 11))
set.seed(8569205)
start_time <- proc.time()['elapsed']

rds_dir <- file.path(config$out_root, 'Rds')
```

# Read data and prep

```{r}
s <- readRDS(file = file.path(rds_dir, '02_integrated_seurat_object.Rds'))
md <- readRDS(file = file.path(rds_dir, '03_integrated_annotated_meta_data.Rds'))
az_umap <- readRDS(file = file.path(rds_dir, '03_integrated_annotated_az_umap.Rds'))
```

Change sample names
```{r}
new_name <- function(old_name) {
  old_name <- stringr::str_replace(old_name, pattern = '_[A-Z0-9]+$', replacement = '')
  ret <- sapply(old_name, function(on) {
    if (grepl(pattern = '^KB[1234]', x = on)) {
      nn <- stringr::str_replace(on, pattern = 'KB(?=[1234])', replacement = 'HC')
    } else {
      n <- as.numeric(gsub('KB', '', on))
      nn <- sprintf('P%d', n - 4)
    }
    return(nn)
  })
  ret
}
md$orig.ident <- new_name(md$orig.ident)
s@meta.data$orig.ident <- new_name(s@meta.data$orig.ident)
s <- AddMetaData(s, md)
```

Set up colors
```{r}
genotype_cols <- c('#8da0cb', '#fc8d62')
genotype_cols <- c('gray80', '#2c7fb8')
genotype_order <- c('WT', 'LTBR mutant')
names(genotype_cols) <- genotype_order
```

# UMAP and composition

Azimuth coordinates and predictions
```{r}
df <- cbind(md, s@reductions$int_umap@cell.embeddings, az_umap@cell.embeddings) %>%
  mutate(UMAP_1 = az_umap_1, UMAP_2 = az_umap_2)
df$genotype <- factor(df$genotype, levels = genotype_order)
```

Group some cell types
```{r}
df <- mutate(df, celltype = case_when(
  stringr::str_starts(az.annotation, 'B ') ~ 'B cells',
  stringr::str_starts(az.annotation, 'NK') ~ 'NK cells',
  stringr::str_starts(az.annotation, 'CD4 ') ~ 'CD4 T cells',
  stringr::str_starts(az.annotation, 'CD8 ') ~ 'CD8 T cells',
  TRUE ~ az.annotation
))
```

Exclude cells with low confidence cell type prediction (use the non-grouped annotation)
```{r}
df_filtered <- filter(df, !is.na(az.annot.filtered))
```

Further exclude cell types we see less than 20 times
```{r}
tab <- table(df_filtered$az.annotation)
keep_ct <- names(tab)[tab >= 20]
df_filtered2 <- filter(df_filtered, az.annotation %in% keep_ct) %>%
  mutate(az.annotation = case_when(
    stringr::str_starts(az.annotation, 'B ') ~ 'B cells',
    stringr::str_starts(az.annotation, 'NK') ~ 'NK cells',
    TRUE ~ az.annotation
  ))

n <- length(unique(df_filtered2$az.annotation))
cell_type_colors <- hcl.colors(n, palette = 'Dark 2')
names(cell_type_colors) <- unique(df_filtered2$az.annotation)
cd8_cell_type_colors <- cell_type_colors[grepl('CD8', names(cell_type_colors))]
```

Create figures
```{r, fig.width=12, fig.height=3, out.width='100%'}
df_lab <- group_by(df_filtered2, az.annotation) %>%
  summarise(UMAP_1 = median(az_umap_1),
            UMAP_2 = median(az_umap_2),
            .groups = 'drop') %>%
  mutate(label = az.annotation)

p4 <- ggplot(df_filtered2, aes(UMAP_1, UMAP_2, color = az.annotation, label = az.annotation)) +
  geom_point(shape = 16, size = 1, alpha = 0.4) + 
  geom_text_repel(data = df_lab, size = 3, color = 'black', max.overlaps = Inf, min.segment.length = 0.2, point.size = NA) +
  theme_classic() +
  theme(legend.position = 'none') +
  theme(axis.ticks = element_blank(), axis.text = element_blank()) +
  scale_color_manual(values = cell_type_colors)

df_lab <- group_by(df_filtered2, az.annotation) %>%
  summarise(UMAP_1 = median(az_umap_1),
            UMAP_2 = median(az_umap_2),
            .groups = 'drop') %>%
  dplyr::rename(label = az.annotation) %>%
  ungroup()
df_lab_empty <- group_by(df_filtered2, az.annotation) %>%
  slice_sample(n = 0) %>%
  mutate(UMAP_1 = az_umap_1,
         UMAP_2 = az_umap_2,
         label = "") %>%
  select(UMAP_1, UMAP_2, label) %>%
  ungroup()


p6 <- slice_sample(df_filtered2, prop = 1) %>%
  ggplot(aes(UMAP_1, UMAP_2)) +
  geom_point(aes(color = genotype), alpha = 1, shape = 19, size = 0.5) + 
  #geom_text_repel(data = bind_rows(df_lab, df_lab_empty), aes(label = label), 
  #                size = 4, label.padding = unit(0.15, 'lines'),
  #                 force = 10, force_pull = 1, min.segment.length = 1) +
  scale_color_manual(values = genotype_cols) +
  theme_classic() +
  theme(axis.ticks = element_blank(), axis.text = element_blank())

cp <- df_filtered2 %>% 
  filter(grepl('CD8', az.annotation)) %>%
  mutate(orig.ident = gsub('N', 'H', orig.ident)) %>%
  composition_plots(group_var = 'orig.ident', comp_var = 'az.annotation', 
                    group_name = 'Sample', comp_name = 'Cell type')
p7 <- cp$fig2 + scale_fill_discrete(name = 'Predicted cell type') + 
  scale_fill_manual(values = cd8_cell_type_colors, name = 'Predicted cell types\nin CD8 compartment') +
  guides(fill=guide_legend(ncol=1)) +
  theme_classic() 

show((p4 + p6 + p7) + plot_layout(nrow = 1, widths = c(2,2,1)))
```

# Fractions of cell types

Fractions for all cells
```{r, fig.width=9, fig.height=5, out.width='100%'}
cp <- composition_plots(df = md, group_var = 'orig.ident', 
                        comp_var = 'az.annotation', group_name = 'Sample', 
                        comp_name = 'Cell type')
p1 <- ggplot(cp$data, aes(Sample, freq)) + geom_point() +
  ylab('Fraction') + xlab('Sample') +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  facet_wrap(~ label_filtered, scales = 'free_y', ncol = 5) +
  theme(legend.position = "none")
show(p1)
```

# Clonotypes

Clonality based on cells with exactly one A and B chain
```{r}
cl_dat <- readRDS(file = file.path(rds_dir, '04_clonotypes.Rds'))
cl_dat$clono_dat$sample_id <- new_name(cl_dat$clono_dat$sample_id)
df <- arrange(cl_dat$clono_dat, sample_id, barcode, chain) %>%
  group_by(sample_id, barcode, chain) %>%
  mutate(chain_n = n(), chain_cdr3 = paste(chain, cdr3, sep=':'), 
         isAorB = stringr::str_starts(chain, 'TR[AB]')) %>%
  ungroup() %>%
  filter(chain_n == 1, isAorB) %>%
  group_by(sample_id, barcode) %>%
  summarise(bc_n = n(), clonotype = paste(chain_cdr3, collapse = ';'), .groups = 'drop') %>%
  filter(bc_n == 2) %>%
  group_by(sample_id) %>%
  summarise(cells = n(), clonotypes = length(unique(clonotype)), clonality = 1 - norm_entropy(clonotype))
df$genotype <- c('LTBR mutant', 'WT')[grepl('^ND', df$sample_id ) + 1]
df$genotype <- factor(df$genotype, levels = c('WT', 'LTBR mutant'))



p1 <- ggplot(df, aes(sample_id, clonality, fill = genotype)) + 
  geom_bar(stat = 'identity') + 
  scale_fill_manual(values = genotype_cols) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  xlab('Sample') + ylab('Clonality') +
  theme(panel.grid.major.y = element_line(rel(0.5), color="grey75")) +
  theme(legend.position = "none")
show(p1)
```


Clonality plots in form of treemap
```{r}
df <- arrange(cl_dat$clono_dat, sample_id, barcode, chain) %>%
  group_by(sample_id, barcode, chain) %>%
  mutate(chain_n = n(), chain_cdr3 = paste(chain, cdr3, sep=':'), 
         isAorB = stringr::str_starts(chain, 'TR[AB]')) %>%
  ungroup() %>%
  filter(chain_n == 1, isAorB) %>%
  group_by(sample_id, barcode) %>%
  summarise(bc_n = n(), clonotype = paste(chain_cdr3, collapse = ';'), .groups = 'drop') %>%
  filter(bc_n == 2) %>%
  group_by(sample_id, clonotype) %>%
  summarise(n = n(), .groups = 'drop')
#install.packages('treemap')

# if we need colors, check, Set2, Set3, Accent
# or just use the genotype colors
fill_col <- colorspace::lighten(genotype_cols, amount = 0.3)

# need to manually arrange treemaps
grid::grid.newpage()
grid::pushViewport(grid::viewport(layout = grid::grid.layout(2, 3), name = 'myvp'))
vplayout <- function(x, y) grid::viewport(layout.pos.row = x, layout.pos.col = y)
i <- 0
for (sam in unique(df$sample_id)) {
  filter(df, sample_id == sam) %>%
    treemap::treemap(index = 'clonotype', vSize = 'n', type = 'index', 
                     palette = fill_col[grepl('^P\\d+$', sam) + 1], 
                     title = sam,
                     vp = vplayout(i %% 2 + 1, i %/% 2 + 1))
  i <- i + 1
}
```

# Differential expression

## Expression levels of DE genes

We want to show CD8 T cells, and their subtypes. Load DE test results.

```{r}
filename <- '05_DE_results.Rds'
de_res <- readRDS(file = file.path(rds_dir, filename))$res %>%
  dplyr::rename(cell_type = split_level) %>%
  filter(grepl(pattern = '^CD8', x = cell_type))
```

Pick which genes to show

```{r}
goi <- filter(de_res, FDR < 0.05) %>%
    group_by(cell_type, sign(logFC)) %>% 
  mutate(FDR_rank = rank(FDR, ties.method = 'min'),
         logFC_rank = rank(-abs(logFC), ties.method = 'min')) %>%
    filter(FDR_rank <= 45 & logFC_rank <= 45) %>%
  pull(gene) %>% unique()
```


Show average expression per sample per cell type for the genes of interest
```{r, fig.width=5, out.width='50%', fig.height=8}
s$grp_id <- paste(s$orig.ident, s$az.annot.filtered)
sample_md <- group_by(s@meta.data, grp_id) %>%
  summarise(genotype = genotype[1], celltype = az.annot.filtered[1], orig.ident = orig.ident[1], n = n(), count_sum = sum(nCount_RNA))

df <- expression_summary(counts = s$RNA@counts, 
                         grouping = factor(s$grp_id), 
                         transformation = 'DESeq2vst', 
                         goi = goi) %>%
  left_join(sample_md) %>%
  filter(!is.na(celltype), grepl(pattern = '^CD8', x = celltype)) %>%
  group_by(gene) %>% 
  #mutate(expr_scaled = scales::rescale(logCPM))
  mutate(expression = scales::oob_squish(x = scale(expression)[, 1], range = c(-2, 2)),
         genotype = stringr::str_replace_all(genotype, 'LTBR mutant', 'M'),
         genotype = factor(genotype, levels = c('M', 'WT'))) %>%
  ungroup()

mat <- acast(df, gene ~ celltype + orig.ident, value.var = 'expression')
tmp <- hclust(dist(mat, method = 'manhattan'), method = 'ward.D2')
df$gene <- factor(df$gene, levels = rownames(mat)[tmp$order])

# add cell number annotation
n_anno <- select(df, orig.ident, celltype, genotype, n) %>%
  unique()

p1 <- ggplot(df, aes(orig.ident, gene, fill = expression)) +
  geom_tile(colour = "gray66") +
  facet_grid(. ~ paste(celltype, genotype), scales = 'free', space = 'free') +
  scale_fill_gradient2(low = '#4d9221', mid = 'white', high = '#c51b7d', midpoint = 0,
                       name = "Expression [rescaled]") +
  geom_text(data = n_anno, aes(label = n, fill = NULL), y = 0, angle = 90, 
            hjust = 1, color = 'gray25', size = 3) +
  theme(panel.spacing = unit(0, "lines")) +
  labs(x = NULL, y = NULL) + 
  coord_cartesian(clip = 'off') +
  scale_x_discrete(expand = c(0, 0)) + 
  scale_y_discrete(expand = c(0, 0)) + 
  theme(axis.text.x=element_blank()) + 
  theme(axis.text.y=element_text(size=rel(0.8))) + 
  theme(strip.text.x.top = element_text(angle = 90, hjust = 0)) +
  theme(legend.position = "top", axis.ticks = element_blank()) +
  theme(plot.margin = margin(t = 1/2, r = 1/2, b = 2, l = 1/2, unit = "line"))
show(p1)
```

## Viz of test results

```{r, fig.width=3, out.width='40%', fig.height=9}
new_goi <- filter(de_res, abs(logFC) > 2 & FDR < 0.05) %>%
  pull(gene) %>%
  unique()
new_goi <- goi

p2 <- de_res %>%
  filter(gene %in% new_goi, FDR < 0.05) %>%
  mutate(gene = factor(gene, levels = levels(df$gene)),
         logFC = scales::oob_squish_any(logFC, c(-3, 3))) %>%
  ggplot(aes(cell_type, gene)) +
  geom_point(aes(size = -log10(FDR), fill = logFC), shape = 21) +
  scale_size_continuous(name = '-log10(FDR)', range = c(1, 5)) +
  scale_fill_gradient2(low = '#e66101', mid = 'white', high = '#5e3c99', midpoint = 0,
                       name = "logFC\nclipped\nto [-3, 3]") +
  theme(panel.spacing = unit(0, "lines")) +
  labs(x = NULL, y = NULL) + 
  theme(axis.ticks = element_blank(), legend.direction = "vertical", legend.box = "horizontal") +
  theme(strip.text.x.top = element_text(angle = 90, hjust = 0)) +
  theme(legend.margin = margin(0,0,0,0),
        legend.position = 'right', axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
show(p2)
```

# Runtime and session info

Runtime: `r time_diff(start_time)`

Session Info
```{r}
sessionInfo()
```
